packages/agents/src/functions/ShellExec.ts

```diff
@@ -0,0 +1,120 @@
+import {
+  Agent,
+  AgentFunctionResult,
+  AgentOutputType,
+  AgentVariables, ChatLogs,
+  ChatMessageBuilder, LlmApi, Tokenizer,
+  trimText
+} from "@evo-ninja/agent-utils";
+import {AgentFunctionBase} from "../AgentFunctionBase";
+import {AgentBaseContext} from "../AgentBase";
+import {FUNCTION_CALL_FAILED} from "../agents/Scripter/utils";
+
+interface ShellExecFuncParameters {
+  command: string;
+  args: string[];
+}
+
+export class ShellExecFunction extends AgentFunctionBase<ShellExecFuncParameters> {
```

Comment:
I'm not a fan of having a generic "shell exec" function right now, it needs some kind of "guard rails" to make sure it doesn't do anything malicious.

Instead I think we should have use-case specific functions, like "run test" which has a known good set of commands it can run (cargo, pytest, yarn/npm, etc).

Basically I think we need to whitelist right now instead of giving the agent open access.

Reply:
let's remove the `ShellExec` function for now, let's only have the exposed `execShell` in the fs workspace and we can implement the `RunPythonTest` on top of this (or any other function, like [poetry init](https://github.com/polywrap/evo.ninja/pull/348))